source/client-backpressure/client-backpressure.md

```diff
@@ -0,0 +1,446 @@
+retryable such as updateMany, create collection, getMore, and generic runCommand. The new command execution method obeys
+the following rules:
+
+1. If the command succeeds on the first attempt, drivers MUST deposit `RETRY_TOKEN_RETURN_RATE` tokens.
+    - The value is 0.1 and non-configurable.
+2. If the command succeeds on a retry attempt, drivers MUST deposit `RETRY_TOKEN_RETURN_RATE`+1 tokens.
+3. If a retry attempt fails with an error that does not include `SystemOverloadedError` label, drivers MUST deposit 1
+    token.
+4. A retry attempt will only be permitted if the error includes the `RetryableError` label, we have not reached
+    `MAX_ATTEMPTS`, the CSOT deadline has not expired, and a token can be acquired from the token bucket.
+    - The value of `MAX_ATTEMPTS` is 5 and non-configurable.
+    - This intentionally changes the behavior of CSOT which otherwise would retry an unlimited number of times within the
+        timeout to avoid retry storms.
+5. If the previous error includes the `SystemOverloadedError` label, the client MUST apply exponential backoff according
+    to according to the following formula: `delayMS = j * min(maxBackoff, baseBackoff * 2^i)`
+    - `i` is the retry attempt (starting with 0 for the first retry).
+    - `j` is a random jitter value between 0 and 1.
+    - `baseBackoff` is constant 100ms.
+    - `maxBackoff` is 10000ms.
+    - This results in delays of 100ms, 200ms, 400ms, 800ms, and 1600ms before accounting for jitter.
+6. If the previous error contained the `SystemOverloadedError` error label, the node will be added to the set of
+    deprioritized servers.
+
+#### Pseudocode
+
+The following pseudocode describes the overload retry policy:
+
+```python
+BASE_BACKOFF = 0.1
+MAX_BACKOFF = 10
+RETRY_TOKEN_RETURN_RATE = 0.1
+
+def execute_command_retryable(command, ...):
+    deprioritized_servers = []
+    attempt = 0
+    while True:
+        try:
+            server = select_server(deprioritized_servers)
+            connection = server.getConnection()
+            res = execute_command(connection, command)
+            # Return tokens to the bucket on success.
+            tokens = RETRY_TOKEN_RETURN_RATE
+            if attempt > 0:
+                tokens += 1
+            token_bucket.deposit(tokens)
+            return res
+        except PyMongoError as exc:
+            backoff = 0
+            attempt += 1
+
+            if attempt > MAX_ATTEMPTS:
+                raise
+
+            # Raise if the error is non retryable.
+            is_retryable = exc.has_error_label("RetryableError") or is_retryable_write_error() or is_retryable_read_error()
+            if not is_retryable:
+                raise error
+            if exc.has_error_label("SystemOverloadedError"):
+                jitter = random.random() # Random float between [0.0, 1.0).
+                backoff = jitter * min(BASE_BACKOFF * (2 ** attempt), MAX_BACKOFF)
```

Comment:
The specification says _"`i` is the retry attempt (starting with 0 for the first retry)."_ The pseudocode deviates from it by using the next attempt number (`attempt`) here instead of using the next retry attempt number (all attempts but the first one are retry attempts). The minimal `attempt` value at this point in execution is 1, making the minimal `backoff` equal to `BASE_BACKOFF * (2^1) = 200 ms` (before accounting for jitter), instead of the expected 100 ms.

To correctly illustrate the specification, the pseudocode should be

```
backoff = jitter * min(BASE_BACKOFF * (2^(attempt - 1)), MAX_BACKOFF)
```

Reply:
Done